Add `HoverBehavior` and utility motion behaviors (`MotionCollideBehavior`, `MotionBlockBehavior`)

DexerBR opened the pull request:
This PR introduces `HoverBehavior`, a new mixin that enables mouse hover detection and specialized hover event handling for Kivy widgets. It also includes two complementary motion event filtering behaviors: `MotionCollideBehavior` and `MotionBlockBehavior`.

## Motivation

Currently, Kivy lacks built-in support for hover events, making it difficult to create responsive UI elements that react to mouse movement. This is especially problematic for:

- Desktop applications requiring hover feedback (tooltips, highlights, previews)
- Widgets in scrollable containers (RecycleView/ScrollView) where hover state needs proper management
- Complex layouts where hover events should be contained to specific areas

The current Kivy recommendation is to bind `Window.mouse_pos` to an update function, but in most cases this is done multiple times in different places. For complex apps, this approach has several drawbacks:

- **Poor maintainability**: Scattered `mouse_pos` bindings across the codebase make hover logic hard to track and debug
- **Performance issues**: Every bound widget receives updates on every mouse movement, even when not needed
- **No collision detection**: Manual `collide_point()` checks must be implemented repeatedly
- **Missing state management**: No built-in tracking of enter/leave events - developers must manually track previous positions
- **Coupling issues**: Widgets become tightly coupled to `Window.mouse_pos`, making testing and reusability difficult
- **Memory leaks**: Forgotten unbindings can cause widgets to persist in memory
- **No event propagation control**: Cannot easily prevent hover "leak-through" to background widgets

`HoverBehavior` solves these issues with a centralized, efficient, and declarative approach that follows Kivy's behavior mixin pattern.

## Changes

### New Behaviors

#### 1. `HoverBehavior` (kivy/uix/behaviors/hover.py)
- **Main feature**: Mixin adding hover detection to any widget
- **Events dispatched**: 
  - `on_hover_enter`: Mouse enters widget bounds
  - `on_hover_update`: Mouse moves within bounds
  - `on_hover_leave`: Mouse exits widget bounds
- **Properties**:
  - `hovered` (read-only): Boolean indicating current hover state
  - `hover_mode`: Controls event propagation ("default", "all", "self")
- **Static hover re-dispatching**: Automatically handles hover when content scrolls under stationary cursor, configurable via `set_hover_update_interval(interval)` method

#### 2. `MotionCollideBehavior` (kivy/uix/behaviors/motion.py)
- Filters motion events to only process those within widget bounds or grabbed
- Essential for RecycleView/ScrollView to prevent hover conflicts outside visible area
- Generic design: works with any motion event type (not just hover)

#### 3. `MotionBlockBehavior` (kivy/uix/behaviors/motion.py)
- Blocks unregistered motion events from passing through widget
- Makes widgets "opaque" to motion events, preventing leak-through to background
- Useful for modal dialogs, overlays, and foreground elements

### Architecture

- **Internal event manager** (`_HoverManager`): Automatically registered on first use
- **Grab-based state tracking**: Maintains hover state across frames
- **Smart leave event synthesis**: Detects when widgets lose hover between frames
- **Window integration**: Seamlessly integrates with Kivy's motion event system

## Usage Examples

### Basic hover with visual feedback
```python
from kivy.uix.behaviors.hover import HoverBehavior
from kivy.uix.label import Label

class HoverLabel(HoverBehavior, Label):
    def on_hovered(self, instance, is_hovered):
        self.color = [1, 0, 0, 1] if is_hovered else [1, 1, 1, 1]
```

### ScrollView with collision filtering
```python
from kivy.uix.behaviors.hover import HoverBehavior
from kivy.uix.behaviors.motion import MotionCollideBehavior
from kivy.uix.scrollview import ScrollView

class HoverScrollView(MotionCollideBehavior, ScrollView):
    pass  # Automatically filters hover outside stencil bounds

class HoverItem(HoverBehavior, MotionCollideBehavior, Label):
    pass  # Only receives hover when visible
```

### Blocking hover leak-through
```python
from kivy.uix.behaviors.motion import MotionBlockBehavior
from kivy.uix.button import Button

class OpaqueButton(MotionBlockBehavior, Button):
    pass  # Blocks unregistered motion events from passing through
```

## Breaking Changes

None. This is a pure addition with no changes to existing APIs.

## Notes

- The hover system is automatically initialized when the first `HoverBehavior` widget is created
- Default hover update interval is 1/30s (configurable globally)
- All three behaviors are designed to be generic and composable
- `MotionCollideBehavior` and `MotionBlockBehavior` should NOT be used together on the same widget


## Context

This PR is **1/3 of the behavior standardization effort** described in https://github.com/kivy/kivy/issues/8521

This PR is based on https://github.com/pythonic64/hover

## Demo video (`.\examples\widgets\hover_behavior.py`)

https://github.com/user-attachments/assets/9d45da63-53d8-44c6-97f0-ae0523acbdcd



<br>

Maintainer merge checklist
* [x] Title is descriptive/clear for inclusion in release notes.
* [x] Applied a `Component: xxx` label.
* [ ] Applied the `api-deprecation` or `api-break` label.
* [ ] Applied the `release-highlight` label to be highlighted in release notes.
* [x] Added to the milestone version it was merged into.
* [ ] **Unittests** are included in PR.
* [ ] Properly documented, including `versionadded`, `versionchanged` as needed.

pythonic64 commented:
@DexerBR Why is my code from https://github.com/pythonic64/hover/ almost entirely copy pasted in this pull request? :D
Why wasn't I tagged on this request or contacted on Discord to make this pull request myself?

Also to add, `EventManager` class, `on_motion` event and other related objects are still in experimental phase which makes code in this request also experimental and my code in `hover` repository is also experimental. IMHO, `hover` should be part of Kivy Garden until we get some feedback on apps actually using it and developers asking for it to be part of the Kivy.

DexerBR commented:
> @DexerBR Why is my code from https://github.com/pythonic64/hover/ almost entirely copy pasted in this pull request? :D Why wasn't I tagged on this request or contacted on Discord to make this pull request myself?
> 
> Also to add, `EventManager` class, `on_motion` event and other related objects are still in experimental phase which makes code in this request also experimental and my code in `hover` repository is also experimental. IMHO, `hover` should be part of Kivy Garden until we get some feedback on apps actually using it and developers asking for it to be part of the Kivy.


Hi @pythonic64! You're right, I should have contacted you first and I apologize for that!

I've been using your implementation for a long time and it has proven to be very stable. I saw mentions of this functionality in past discussions like in this [issue](https://github.com/kivy/kivy/issues/8257), please see after https://github.com/kivy/kivy/issues/8257#issuecomment-1586258969, but those discussions died out.

I believe this is a very important component to have in Kivy `3.x.x` and would be well received by the community, which is why I decided to move forward with the PR.

I added a reference at the top of the PR: "**This PR is based on https://github.com/pythonic64/hover**". **I didn't simply copy-paste your code** - I extracted the core logic and refactored it entirely for better maintainability, along with other modifications I deemed necessary for integration into Kivy core. Both projects are under MIT license.

That said, if you'd prefer to take ownership of this, feel free to tag @misl6 here and request a revert so you can remake it yourself, or you can make a PR on top of this one to change whatever you think is necessary or completely remove it from kivy `3.x.x` (though I believe this feature would be valuable for the community).

I'm not looking to fight over code! My goal has always been to contribute to making Kivy `3.x.x` better and more complete, and I genuinely believe this feature is important for the community.

Best regards!

pythonic64 commented on examples/widgets/hover_behavior.py:
```diff
@@ -0,0 +1,188 @@
+        super().__init__(**kwargs)
+        self.original_color = self.background_color[:]
+
+    def on_state(self, _, state):
```
Property `background_color` depends on `state` and `hovered` properties and it this case seems better to be defined as `AliasProperty` with a getter which calculates the correct color.

pythonic64 commented on examples/widgets/hover_behavior.py:
```diff
@@ -0,0 +1,188 @@
+                [0.8, 0, 0.8, 1] if self.hovered else [0.2, 0.2, 0.2, 1]
+            )
+
+    def on_hover_enter(self, me):
```
Generally you don't need to override this method to handle a hover event, unless you want to inspect `me` itself. Methods `*_enter` and `*_leave` are called every time a hover event enters or leaves this widget which doesn't have to be a hover from the mouse, but from any hover event (laser pointer or anything dispatching hover events).

To just know if this widget is hovered, you only need to "look at" `hovered` property.

pythonic64 commented on examples/widgets/hover_behavior.py:
```diff
@@ -0,0 +1,188 @@
+    def refresh_view_attrs(self, rv, index, data):
+        self.index = index
+        self.base_text = data.get("text", "")
+        return super().refresh_view_attrs(rv, index, data)
```
Always refresh `bg_color` by inspecting `hovered` property in this method.
